libraries/WiFi/src/ETH.cpp

```diff
@@ -385,8 +385,16 @@ bool ETHClass::config(IPAddress local_ip, IPAddress gateway, IPAddress subnet, I
         info.gw.addr = 0;
         info.netmask.addr = 0;
 	}
+    
+    // avoid error messages or failure while DHCP still did not get stopped
+    uint8_t tries = 5;
+    do {
+      const TickType_t xDelay = 50 / portTICK_PERIOD_MS;
+      vTaskDelay( xDelay );
```

Comment:
> since this is Arduino code, why not use `delay(50);` here instead?

Yes, both work the same way. I just used FreeRTOS to do it. 
Let me know if you need me to use ``` delay(50); ``` instead.

Reply:
yes, use `delay()` in arduino code, where you deal with milliseconds. Use vTaskDelay only to delay RTOS ticks

Reply:
as for implementation, it kinda feels a bit hackish :) what will happen if you check if dhcp is started and if so, then turn it off and wait for it to be off. I guess we do not have access to DHCP events?

Reply:
It's not hackish. 
It's a fact that when ETH.begin() is called, a lot of processes take place. 
SPI is used to receive and send TCP/IP packages from the ETH card in the network. 
DHCP server is contacted to obtain an IP address and DHCP packages start to be received along with Arduino Sketch execution.
It's necessary to stop the DHCP process in a graceful way before setting an static IP address in order to avoid getting error messages and sometimes an error when setting the static IP.
Arduino layer has no access to DHCP events. It's in the LWIP level within IDF. 

Arduino only gets ```ARDUINO_EVENT_ETH_GOT_IP``` in WiFi.onEvent() - which is funny because WiFi and Ethernet are mixed all together in the ETH implementation. This event is received for both DHCP and Static IP. But we must think that a DHCP server is not mandatory in a network, thus it can also just get timed out.

Thus the solution is not so simple. 
This PR solves the issue in a transparent way for a whole set of networking situatuions.

Reply:
> yes, use `delay()` in arduino code, where you deal with milliseconds. Use vTaskDelay only to delay RTOS ticks

OK. No problem. As I said both are exactly the same as seen in 
https://github.com/espressif/arduino-esp32/blob/master/cores/esp32/esp32-hal-misc.c#L173-L176

Reply:
I changed it to use ```delay(50);``` as you have requested.

Reply:
@me-no-dev 
I tested it further. Everything lays down to just a ```delay(50)``` in order to force FreeRTOS to change context and allow DHCP to process its events getting into a stable state in order to set static IP with no error messages.
Thus the "hack" you mentioned is now gone.

Reply:
I will suggest that you first check if DHCP is even started: https://github.com/espressif/arduino-esp32/blob/371f382db7dd36c470bb2669b222adf0a497600d/tools/sdk/esp32/include/tcpip_adapter/include/tcpip_adapter.h#L164 statuses are defined here: https://github.com/espressif/arduino-esp32/blob/4413dbbd8746f48201c2699702a7a5f13cf9ac79/tools/sdk/esp32s2/include/esp_netif/include/esp_netif_types.h#L61

Reply:
on the `delay` use: yes it is in fact the same thing, but since it's Arduino we are talking about, it's better to use it's API. That way it will also be more clear to others reading the code.

Reply:
I tested ```tcpip_adapter_dhcps_get_status()``` and it never returns ESP_OK. 
It returns ```  [  4769][E][ETH.cpp:394] config(): Error (258) getting DHCP Status (5) ```
It is the same with or without a DHCP server in the network.

Thus it is not possible to use it to check if DHCP has started using this API.

Here is the code I used:

``` java

    tcpip_adapter_dhcp_status_t status;
    err = tcpip_adapter_dhcps_get_status(TCPIP_ADAPTER_IF_ETH, &status);
    if (err == ESP_OK) {
      log_e("DHCP Status = %d\n", status);
    } else {
      log_e("Erro r(%d) getting DHCP Status (%d)\n", err, status);
    }


```

Reply:
So far, the solution with ```delay(50);``` is the only one that works.

Reply:
maybe we should report this to IDF? we can go with 50ms but that is what I meant by "hackish" it's an arbitrary delay, not something precise. It is possible to end up in situation that it will not be enough. Maybe SPI ETH chips will require this to be different (especially the 10Mbps ones)

Reply:
What would happen if we delay in `begin` instead?

Reply:
> What would happen if we delay in `begin` instead?

It works in the same way. That's the solution I proposed in the issue #5733

Reply:
> maybe we should report this to IDF? we can go with 50ms but that is what I meant by "hackish" it's an arbitrary delay, not something precise. It is possible to end up in situation that it will not be enough. Maybe SPI ETH chips will require this to be different (especially the 10Mbps ones)

I agree to you. That's why I had this "4 tries of 50ms" previously in the code.

Reply:
I think this issue is like that - out of our control because it depends on the timing of an external DHCP server.
Other than that it must be fixed in IDF. But there was already some people complaining about it and the answer was something like "ignore it!".

https://www.esp32.com/viewtopic.php?t=6922
https://github.com/espressif/esp-mdf/issues/6

Reply:
> > What would happen if we delay in `begin` instead?
> 
> It works in the same way. That's the solution I proposed in the issue #5733

OK, let's move it there, so you do not need to delay if you change previously assigned static IP (and DHCP is already off)

Reply:
> > > What would happen if we delay in `begin` instead?
> > 
> > 
> > It works in the same way. That's the solution I proposed in the issue #5733
> 
> OK, let's move it there, so you do not need to delay if you change previously assigned static IP (and DHCP is already off)

ETH doesn't allow to assign static IP before ```ETH.begin()```. If we do so, it thrown an exception and resets.
Booting up ETH with ```begin()``` starts automatically DHCP.

Reply:
yes yes, that is why I am saying to add it to `begin` so it waits for DHCP to start. Then you can set the static IP as many times as you want without delay

Reply:
So you are saying to do not change ```ETH.config()``` and just advise users about it, thus leave it this way and do not apply this PR?

Reply:
no no :) add the code at the end of `ETH.begin`. Since that is the code that starts DHCP it will start it and wait there 50ms, then when the user calls config, dhcp will already be in state that can be stopped.

Reply:
Done. It did work fine and error messages are gone. 
Changes made to ```ETH.begin```